Userland/Libraries/LibPDF/Renderer.cpp

```diff
@@ -299,16 +308,103 @@ RENDERER_HANDLER(path_append_rect)
+    m_clipped_bitmap_painter->add_clip_rect(r);
+
+    // Need to copy what's currently on the page so that transparent things in the clipped area are blended on top of the right things.
+    m_clipped_bitmap_painter->blit(r.location(), *m_bitmap, r);
+
+    return {};
+}
+
+static void apply_clip(Gfx::Bitmap& bitmap, Gfx::IntRect bitmap_rect, Gfx::Bitmap const& clip_alpha, Gfx::IntRect clip_rect)
+{
+    VERIFY(bitmap.size() == bitmap_rect.size() || (bitmap_rect.is_empty() && bitmap.size() == Gfx::IntSize(1, 1)));
+    VERIFY(clip_alpha.size() == clip_rect.size() || (clip_rect.is_empty() && clip_alpha.size() == Gfx::IntSize(1, 1)));
+    auto rect = bitmap_rect.intersected(clip_rect);
+
+    for (int y = rect.top(); y < rect.bottom(); ++y) {
+        for (int x = rect.left(); x < rect.right(); ++x) {
+            Gfx::IntPoint point { x, y };
+            auto pixel = bitmap.get_pixel(point - bitmap_rect.location());
+            auto alpha = clip_alpha.get_pixel(point - clip_rect.location());
+            pixel.set_alpha(pixel.alpha() * alpha.alpha() / 255);
+            bitmap.set_pixel(point - bitmap_rect.location(), pixel);
+        }
+    }
+}
+
+void Renderer::copy_current_clip_path_content_to_output()
+{
+    if (state().clipping_state.clip_path_alpha) {
+        auto r = state().clipping_state.clip_bounding_box;
+        apply_clip(*m_clipped_bitmap, m_clipped_bitmap->rect(), *state().clipping_state.clip_path_alpha, r);
+        m_painter.blit(r.location(), *m_clipped_bitmap, r);
+    }
+
+    m_clipped_bitmap_painter = nullptr;
+    m_clipped_bitmap_anti_aliasing_painter = nullptr;
+}
+
+PDFErrorOr<void> Renderer::add_clip_path(Gfx::WindingRule winding_rule)
+{
+    if (m_rendering_preferences.show_clipping_paths)
+        m_clip_paths_to_show_for_debugging.append(m_current_path);
+
+    if (!m_rendering_preferences.apply_clip)
+        return {};
+
+    // If the clip is a rectangle that is larger than the current clip's bounding box, we can ignore it.
+    // This is important for performance, because many PDFs add a clip for the whole page before other clips,
+    // and we don't want to allocate a whole-page bitmap for those.
+    if (auto maybe_rect = m_current_path.as_rect(); maybe_rect.has_value()) {
+        auto rect = Gfx::enclosing_int_rect(maybe_rect.value());
+        if (rect.contains(state().clipping_state.clip_bounding_box))
+            return {};
+    }
+
+    // About to set new clip; flush old one if needed.
+    copy_current_clip_path_content_to_output();
+
+    auto next_clipping_bbox = Gfx::enclosing_int_rect(m_current_path.bounding_box());
+    next_clipping_bbox.intersect(state().clipping_state.clip_bounding_box);
+
+    auto clip_path_alpha = MUST(Gfx::Bitmap::create(Gfx::BitmapFormat::BGRA8888, next_clipping_bbox.is_empty() ? Gfx::IntSize { 1, 1 } : next_clipping_bbox.size()));
+
+    m_current_path.close_all_subpaths();
+    Gfx::Painter clip_painter(*clip_path_alpha);
```

Comment:
Is there a reason not to use the AA painter here? It'd be consistent with the other uses of `.fill_path()` in LibPDF and given the alpha is the only thing you care about here it probably doesn't hurt to bump the samples a little.

Reply:
Good call, just forgot it. Done, I think. Hope I got the API right.